### ISSUE TITLE
Validation data aggregates into training data for multi-drug models

### ISSUE THREAD

**Issue by nictru** [daisybio/drevalpy#349]
I am not fully sure about the consequences of this bug yet (still running the benchmark with a fixed setup). Still I already wanted to share this:

My setup (the DrugANVAEModel is a custom model I am developing, but this does not matter too much here):

```python
from drevalpy.experiment import drug_response_experiment
from drevalpy.visualization.create_report import create_report
from drevalpy.models import MODEL_FACTORY
from drevalpy.datasets import AVAILABLE_DATASETS

import torch

torch.set_float32_matmul_precision('medium')

dataset = "CTRPv2"
run_id = "drugANVI"
test_mode = "LCO"
n_cv_splits = 5
hyperparameter_tuning = False
path_data = "data"
path_out = "results"

baselines = [
    'NaivePredictor',
    'NaiveTissueMeanPredictor',
    'NaiveTissueDrugMeanPredictor',
]

models = [
    'SimpleNeuralNetwork',
    'ElasticNetModel',
    'PCAElasticNet',
    'ScVIElasticNet',
    'RandomForest',
    'PCARandomForest',
    'ScVIRandomForest',
    'DrugANVAEModel'
]

drug_response_experiment(
            models=[MODEL_FACTORY[model] for model in models],
            baselines=[MODEL_FACTORY[baseline] for baseline in baselines],
            response_data=AVAILABLE_DATASETS[dataset](path_data="data"),
            n_cv_splits=n_cv_splits,
            test_mode=test_mode,
            run_id=run_id,
            path_data=path_data,
            path_out=path_out,
            hyperparameter_tuning=hyperparameter_tuning,
            wandb_project=run_id
)

create_report(run_id=run_id, dataset=dataset)
```

What I did:

1. Run the experiment 'from scratch'
2. Delete the DrugANVAEModel from the results directory
3. Re-run the script (all models except DrugANVAEModel are cached)

What I observe:

During the initial training (from scratch), the DrugANVAEModel trains for 500+ steps per epoch, in the second training it trains only for 300+ steps per epoch. Also the convergence in the initial run is way smoother and reaches a loss of 0.7, whereas in the second run the best I get is 1.7 with very shaky convergence. (I made this observations using weights and biases).

Investigating potential reasons:

It seems like [`get_datasets_from_cv_split()`](https://github.com/daisybio/drevalpy/blob/587cefd8e3c60151e85ee311d2f36d7e7b5a4a47/drevalpy/experiment.py#L1259) returns direct references instead of copies for multi-drug models. This is problematic as [here](https://github.com/daisybio/drevalpy/blob/587cefd8e3c60151e85ee311d2f36d7e7b5a4a47/drevalpy/experiment.py#L224), the validation data is added to the training data. Now when training many models, the validation data is actually added to the training data multiple times, resulting in my observed 500+ steps per epoch compared to 300+ when only one model is actually trained (and the rest is cached).

**Comment by nictru** [comment]
Here is how the wandb plots compare:

<img width="368" height="629" alt="Image" src="https://github.com/user-attachments/assets/7d68f0c5-16de-493d-bdcc-4e9c4833f46b" />

**Comment by PascalIversen** [comment]
Oh no ... thanks for finding this!!! Sorry for that oversight...

@JudithBernett, I hope and think that the NF pipeline (and therefore our benchmark) is not affected, since we just load the pickled data there?

Fix by just copying here, right?

```python
def get_datasets_from_cv_split(
    split: dict[str, DrugResponseDataset], model_class: type[DRPModel], model_name: str, drug_id: str | None = None
) -> tuple[DrugResponseDataset, DrugResponseDataset, DrugResponseDataset | None, DrugResponseDataset]:
    """
    Get train, validation, (early stopping), and test datasets from the CV split.

    :param split: dictionary of the CV split
    :param model_class: model class
    :param model_name: model name
    :param drug_id: drug id for single drug models
    :returns: tuple of train, validation, (early stopping), and test datasets
    """
    train_dataset = split["train"].copy() 
    validation_dataset = split["validation"].copy()
    test_dataset = split["test"].copy()
```

**Comment by nictru** [comment]
Yes, I fixed it like you pointed out. I am currently running my training setup again to validate if the number of steps per epoch is now correct.

---

Also, another small thing (not even sure if it has any effect):
When doing the full run from scratch, the `tissue` column was added to the per-split prediction files. In the second run's predictions, this column was absent. Most likely this is because the tissue column is not specified [here](https://github.com/daisybio/drevalpy/blob/587cefd8e3c60151e85ee311d2f36d7e7b5a4a47/drevalpy/datasets/dataset.py#L478). 
I am not sure if this can potentially cause issues for certain models, for mine it actually did not have any effect as I don't use the tissue information.

```python
tr_split = DrugResponseDataset.from_csv(os.path.join(path, split_train), dataset_name=self.dataset_name)
te_split = DrugResponseDataset.from_csv(os.path.join(path, split_test), dataset_name=self.dataset_name)
```

**Comment by JudithBernett** [comment]
Ohh I see - this is indeed not good :D Thanks for spotting+fixing it!
The pipeline and the benchmark are not affected because there, we load the pickled, that's right :) So all good for the paper!

The pipeline produces the split pkls: 
https://github.com/nf-core/drugresponseeval/blob/dev/modules/local/cv_split/main.nf

and then each combination of model_name + test_mode + hyperparameter_configuration calls on get_datasets_from_cv_split.
https://github.com/nf-core/drugresponseeval/blob/dev/modules/local/train_and_predict_cv/main.nf

The other bug is a good point, too :) The tissue column was introduced later; we probably forgot this here. We should just fix it here: https://github.com/daisybio/drevalpy/blob/587cefd8e3c60151e85ee311d2f36d7e7b5a4a47/drevalpy/datasets/dataset.py#L87-93

by writing
`tissue_column: str = "tissue"`

because we anyway check whether the column exists
https://github.com/daisybio/drevalpy/blob/587cefd8e3c60151e85ee311d2f36d7e7b5a4a47/drevalpy/datasets/dataset.py#L126-133

**Comment by nictru** [comment]
In the meantime, my re-evaluation with the fix is finished. The number of steps per epoch is correct now. Also, it makes my model perform worse in the report (it used to be better than the random forest, now it is worse) :/

I'm just wondering what the actual reason for my model becoming worse in the report can be:
Before fixing the bug, the model was trained on validation+training data, and then obviously performed good on the validation data. But why did it also perform better on the test data?

Is the validation data also the test data (if hyperparameter tuning is disabled) and thus my model already saw the test data during training? 

Or did the test data _also_ leak into the training data?

**Comment by PascalIversen** [comment]
Did you use the built-in early stopping for your model? It could be that some training data leaked into the early stopping data through the bug, which caused your model to train longer, improving performance by coincidence (the early stopping is strict, since it relies on leave cell line out stricts, which can stop some models quite early)

Or the validation data from previous splits leaked into the current train? This would be test-leakage and definetly explain it...


If hpam tuning is disabled, the model should go directly to the final model training stage, where it's trained on train+validation and predicts test. I don't think we ever touch the test data.



To be safer, I think we should also add copy() here to the datasets, so the models can't manipulate the datasets:


```python
            model.train(
                output=train_dataset,
                output_earlystopping=early_stopping_dataset,
                cell_line_input=cl_features.copy(),
                drug_input=drug_input,
                model_checkpoint_dir=temp_dir,
            )

```

also here for the cell line and drug ids, which are also mutable.

```python
        prediction_dataset._predictions = model.predict(
            cell_line_ids=prediction_dataset.cell_line_ids,
            drug_ids=prediction_dataset.drug_ids,
            cell_line_input=cl_features.copy(),
            drug_input=drug_input,
        )
```

Or we just copy at the beginning of train_and_predict?
